### PR TITLE
Adds dialyzer support

### PR DESCRIPTION
**Pull request by sobolevn**
That's not working yet. Please do not merge.

Here are the steps to reproduce:
1. `mix deps.get`
2. `mix dialyzer`

Errors:

> Starting Dialyzer
> dialyzer args: [check_plt: false,
>  init_plt: '/Users/sobolev/Documents/github/recaptcha/_build/dev/dialyxir_erlang-19.3_elixir-1.4.4_deps-dev.plt',
>  files_rec: ['/Users/sobolev/Documents/github/recaptcha/_build/dev/lib/recaptcha/ebin'],
>  warnings: [:unknown]]
> done in 0m2.2s
> :0: Unknown function 'Elixir.EEx.Engine':'fetch_assign!'/2
> :0: Unknown function 'Elixir.Poison':decode/1
> lib/recaptcha.ex:28: Function verify/1 has no local return
> lib/recaptcha.ex:28: Function verify/2 has no local return
> lib/recaptcha/http.ex:39: Function request_verification/1 has no local return
> lib/recaptcha/http.ex:39: Function request_verification/2 has no local return
> lib/recaptcha/http.ex:40: The call 'Elixir.Recaptcha.Config':get_env('recaptcha','timeout',5000) breaks the contract (atom(),atom(),atom() | map()) -> term()
> lib/recaptcha/http.ex:41: The call 'Elixir.Recaptcha.Config':get_env('recaptcha','verify_url',<<_:376>>) breaks the contract (atom(),atom(),atom() | map()) -> term()
> done (warnings were emitted)

### CODE REVIEW

**Comment by JustMikey** [comment]
The spec is wrong [here](https://github.com/samueljseay/recaptcha/blob/0615b46466cece8ce2160cc5093365f6dfe2081e/lib/recaptcha/config.ex?L9#L9), the third parameter probably should be `any`, this may remove a couple of warnings. I will look into it once I have some free time.

**Comment by JustMikey** [comment]
`has no local return` warning usually means that the spec is wrong somewhere or that the function will actually raise OR can raise and in this case, spec should have the `no_return` option in it to show that raising is expected, especially for the bang functions.

**Comment by sobolevn** [comment]
Fixed almost everything (thanks @JustMikey), but it is still failing. No idea why.

Logs:
> Starting Dialyzer
> dialyzer args: [check_plt: false,
>  init_plt: '/Users/sobolev/Documents/github/recaptcha/_build/dev/dialyxir_erlang-19.3_elixir-1.4.4_deps-dev.plt',
>  files_rec: ['/Users/sobolev/Documents/github/recaptcha/_build/dev/lib/recaptcha/ebin'],
>  warnings: [:unknown]]
> done in 0m4.08s
> :0: Unknown function 'Elixir.EEx.Engine':'fetch_assign!'/2
> :0: Unknown function 'Elixir.Poison':decode/1
> done (warnings were emitted)

**Comment by sobolevn** [comment]
I am done for today, status:

> Starting Dialyzer
> dialyzer args: [check_plt: false,
>  init_plt: '/Users/sobolev/Documents/github/recaptcha/_build/dev/dialyxir_erlang-19.3_elixir-1.4.4_deps-dev.plt',
>  files_rec: ['/Users/sobolev/Documents/github/recaptcha/_build/dev/lib/recaptcha/ebin'],
>  warnings: [:unknown]]
> done in 0m1.55s
> lib/recaptcha/template.ex:20: Function display/0 has no local return
> lib/recaptcha/template.ex:20: Function display/1 has no local return
> lib/recaptcha/template.ex:22: The call 'Elixir.Recaptcha.Template':render_template([{'options','nil' | [{atom(),_}] | map()} | {'public_key',_},...]) will never return since it differs in the 1st argument from the success typing arguments: (map())
> lib/template.html.eex:4: The call 'Elixir.EEx.Engine':'fetch_assign!'(assigns@1::[{'options','nil' | [any()] | map()} | {'public_key',_},...],'public_key') breaks the contract (map(),'Elixir.Map':key()) -> term() | 'nil'
> lib/template.html.eex:13: Function render_template/1 has no local return
> done (warnings were emitted)

**Comment by JustMikey** [comment]
A few things to notice, even though not exactly concerning the specs here:

1) Elixir from version 1.4 I believe starts to complain about not having `()` after function calls so it may make sense to add those everywhere.
2) `String.to_existing_atom()` is a good idea but it may raise if the atom is not in the system yet. So if you just start an application and it doesn't have the recaptcha error atoms in its atoms table, it is gonna raise so you need to make sure that those atoms exist right after the application start.

File: mix.exs [hunk 1]
```diff
@@ -39,8 +45,10 @@ defmodule Recaptcha.Mixfile do
     [
       {:httpoison, "~> 0.11.0"},
       {:poison, "~> 1.5 or ~> 2.0 or ~> 3.0"},
+
```
Comment (samueljseay) [inline mix.exs]: Why the blank line here

File: README.md [hunk 2]
```diff
@@ -76,7 +76,7 @@ Recaptcha provides the `verify/2` method. Below is an example using a Phoenix co
 
 ```elixir
   def create(conn, params) do
-    # some code  
+    # some code
```
Comment (samueljseay) [inline README.md]: Maybe this comment can just be removed? I don't think it adds anything

File: mix.exs (hunk 1 above)
Comment (sobolevn) [inline mix.exs]: This line serves me quite well to separate actual requirements from the development packages.

Comment (samueljseay) [inline mix.exs]: alright 😄 I don't mind

**Comment by sobolevn** [comment]
Done, it works at last.

> » mix dialyzer
> Compiling 1 file (.ex)
> Checking PLT...
> [:eex, :elixir, :httpoison, :kernel, :logger, :poison, :stdlib]
> PLT is up to date!
> Starting Dialyzer
> dialyzer args: [check_plt: false,
>  init_plt: '/Users/sobolev/Documents/github/recaptcha/_build/dev/dialyxir_erlang-19.3_elixir-1.4.4_deps-dev.plt',
>  files_rec: ['/Users/sobolev/Documents/github/recaptcha/_build/dev/lib/recaptcha/ebin'],
>  warnings: [:unknown]]
> done in 0m1.8s
> done (passed successfully)

**Comment by samueljseay** [comment]
@JustMikey do you have a reference for 1.4 expecting `()` I've been out of the Elixir game for a few months

**Comment by sobolevn** [comment]
@samueljseay https://groups.google.com/forum/#!topic/elixir-lang-core/Otz0uuML764

**Comment by samueljseay** [comment]
Thanks @sobolevn I'm sure I or someone else could submit a follow up PR at some point to fix this warning. I don't mind if it doesn't get addressed here.

**Comment by sobolevn** [comment]
Squashed, renamed! Any other things to do before merging it?

**Comment by JustMikey** [comment]
@sobolevn I am still bothered with this `to_existing_atom` call though, have you tested it with all the possible errors from the google?

**Comment by sobolevn** [comment]
@JustMikey that's good point! Nope, I did not test it with every possible error.
Any ideas on how to do that?

Maybe `to_atom` is fine?

**Comment by JustMikey** [comment]
@sobolevn Well, I am not sure here. I think they have all the possible errors on their page and there are like 4 of them so you could generate those atoms in advance and leave the `to_existing_atom` call or just return to using the `to_atom` call, hopefully Google can be trusted not to generate errors dynamically :)

**Comment by sobolevn** [comment]
`to_atom` is easier to understand, support, and implement.  Changed this line, added comment linking to this discussion. Thank you for guidance.

**Comment by JustMikey** [comment]
Of course. Thank you for all the hard work :)

**Pull request merged**